// @dyteinternals/callstats/dist/types.js
export var BrowserTypes;
(function (BrowserTypes) {
    BrowserTypes["CHROMIUM"] = "chromum";
    BrowserTypes["FIREFOX"] = "firefox";
    BrowserTypes["SAFARI"] = "safari";
})(BrowserTypes || (BrowserTypes = {}));
export var ClientEnvType;
(function (ClientEnvType) {
    ClientEnvType["DEV"] = "dev";
    ClientEnvType["STAGING"] = "staging";
    ClientEnvType["PRODUCTION"] = "production";
})(ClientEnvType || (ClientEnvType = {}));
export var MediaDeviceType;
(function (MediaDeviceType) {
    MediaDeviceType["AUDIO"] = "AUDIO";
    MediaDeviceType["VIDEO"] = "VIDEO";
    MediaDeviceType["SPEAKER"] = "SPEAKER";
})(MediaDeviceType || (MediaDeviceType = {}));
export var PermissionEnum;
(function (PermissionEnum) {
    PermissionEnum[PermissionEnum["INIT"] = 0] = "INIT";
    PermissionEnum[PermissionEnum["ACCEPTED"] = 1] = "ACCEPTED";
    PermissionEnum[PermissionEnum["DENIED"] = 2] = "DENIED";
    PermissionEnum[PermissionEnum["SYS_DENIED"] = 3] = "SYS_DENIED";
    PermissionEnum[PermissionEnum["FAILED"] = 4] = "FAILED";
    PermissionEnum[PermissionEnum["NOTFOUND"] = 5] = "NOTFOUND";
    PermissionEnum[PermissionEnum["NOT_APPLICABLE"] = 6] = "NOT_APPLICABLE";
})(PermissionEnum || (PermissionEnum = {}));
